### PR TITLE
Annotations

### PR DESCRIPTION
**Pull request by barnjamin**
This PR adds the ability to specify `Annotated` parameters.

The only annotation allowed currently is a `ParameterAnnotation` which takes 2 arguments:

- `descr`: the string description to add to the contract definition
- `default`: The arguments default value that can be resolved (see DefaultArgument)

Usage:

```py
    @external
    def hash_it(
        self,
        input: Annotated[abi.String, ParameterAnnotation(descr="The input to hash")],
        iters: Annotated[
            abi.Uint64, ParameterAnnotation(descr="The number of times to iterate")
        ],
        opup_app: Annotated[
            abi.Application,
            ParameterAnnotation(
                descr="The app id to use for opup reququests",
                default=OpUp.opup_app_id,
            ),
        ],
        *,
        output: abi.StaticArray[abi.Byte, Literal[32]],
    ):
        return Seq(
            Assert(opup_app.application_id() == self.opup_app_id),
            self.call_opup(Int(255)),
            (current := ScratchVar()).store(input.get()),
            For(
                (i := ScratchVar()).store(Int(0)),
                i.load() < iters.get(),
                i.store(i.load() + Int(1)),
            ).Do(current.store(Sha256(current.load()))),
            output.decode(current.load()),
        )
   ```
   
Results in the following ABI Method
```json

    {
      "name": "hash_it",
      "args": [
        {
          "type": "string",
          "name": "input",
          "desc": "The input to hash"
        },
        {
          "type": "uint64",
          "name": "iters",
          "desc": "The number of times to iterate"
        },
        {
          "type": "application",
          "name": "opup_app",
          "desc": "The app id to use for opup reququests"
        }
      ],
      "returns": {
        "type": "byte[32]"
      }
    }

   ```

### CODE REVIEW

File: examples/opup/contract.py [hunk 1]
```diff
@@ -1,30 +1,84 @@
...
-    @handler(resolvable=ResolvableArguments(opup_app=OpUp.opup_app_id))
+    @handler
     def hash_it(
         self,
-        input: abi.String,
-        iters: abi.Uint64,
-        opup_app: abi.Application,
+        input: Checked[abi.String, nonzero()],
+        iters: Checked[abi.Uint64, nonzero()],
+        _: Checked[abi.Application, equal(OpUp.opup_app_id)],
```
Comment (deanstef) [inline examples/opup/contract.py:63]: Got you. I like the way you are using Annotation here instead of `ResolvableArguments`

**Review by deanstef — APPROVED** [review]
I like your use of `Annotated`. It gives more expressiveness to method calls. We can use it for:
- Arguments descriptions;
- Resolvable arguments / default values;
- Exec custom logic (probably an example is needed for that). 

LGTM

**Comment by cusma** [comment]
I prefer this explicit `Annotation` rather than the automatic resolvable behavior! For example there could be methods with `OpUp` for which is useful to know in advance how much fees I should pay for the execution of that method, I think is good to annotate that in the method's description.

**Comment by barnjamin** [comment]
Good point on fees required. I'm not sure that will be easy to expose until we have some better way to track costs programmatically 🤔

**Pull request merged**